### PR TITLE
feat(coderd): add DERP healthcheck

### PR DESCRIPTION
**Pull request by coadler**
Updates: https://github.com/coder/coder/issues/6728

### CODE REVIEW

File: coderd/healthcheck/accessurl.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func (r *AccessURLReport) Run(ctx context.Context, accessURL *url.URL) {
+	accessURL, err := accessURL.Parse("/healthz")
+	if err != nil {
+		r.Err = xerrors.Errorf("parse healthz endpoint: %w", err)
+		return
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", accessURL.String(), nil)
```
Comment (deansheather) [inline coderd/healthcheck/accessurl.go:26]: Even though the healthcheck has a 30 second context, ideally the access URL healthcheck should fail if it takes more than 5 seconds or so.

File: coderd/healthcheck/healthcheck.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+	"tailscale.com/tailcfg"
+)
+
+type Report struct {
```
Comment (deansheather) [inline coderd/healthcheck/healthcheck.go:11]: Ideally there's an overall "Pass" or "Fail" grade, as well as a "Pass" or "Fail" for each of the sub-reports.

File: coderd/healthcheck/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	Time time.Time  `json:"time"`
+	DERP DERPReport `json:"derp"`
+
+	// TODO
+	// AccessURL AccessURLReport
```
Comment (deansheather) [inline coderd/healthcheck/healthcheck.go:16]:
```suggestion
	AccessURL AccessURLReport
```

File: coderd/healthcheck/healthcheck.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+type ReportOptions struct {
+	// TODO: support getting this over HTTP?
+	DERPMap *tailcfg.DERPMap
+}
+
+func Run(ctx context.Context, opts *ReportOptions) (*Report, error) {
+	var report Report
+
+	err := report.DERP.Run(ctx, &DERPReportOptions{
```
Comment (deansheather) [inline coderd/healthcheck/healthcheck.go:28]: access url report

File: coderd/healthcheck/derp.go [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+	r.mu.Unlock()
+
+	client, err := derphttp.NewClient(key.NewNode(), derpURL.String(), func(format string, args ...any) {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+
+		msg := fmt.Sprintf(format, args...)
+		if strings.Contains(msg, "We'll use WebSockets on the next connection attempt") {
+			r.UsesWebsocket = true
+		}
```
Comment (deansheather) [inline coderd/healthcheck/derp.go:335]: nice touch

**Review by deansheather — APPROVED** [review]

File: coderd/healthcheck/healthcheck.go (hunk 2 above)
Comment (coadler) [reply coderd/healthcheck/healthcheck.go:11]: I thought about that but at least for DERP pass/fail felt more like a spectrum. You could have a derp server that could exchange messages but was listed as STUN but STUN wasn't working, etc

File: coderd/healthcheck/accessurl.go (hunk 1 above)
Comment (coadler) [reply coderd/healthcheck/accessurl.go]: I forgot I wrote this, gonna leave out of this pr

File: coderd/healthcheck/healthcheck.go (hunk 3 above)
Comment (coadler) [reply coderd/healthcheck/healthcheck.go:16]: same as above

File: coderd/healthcheck/healthcheck.go (hunk 2 above)
Comment (coadler) [reply coderd/healthcheck/healthcheck.go:11]: Eh maybe "Fail" could just mean anything failed or is degraded.

**Pull request merged**